src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs

```diff
@@ -50,43 +64,106 @@ internal class DuplexPipeStreamAdapter<TStream> : DuplexPipeStream, IDuplexPipe
 
         public TStream Stream { get; }
 
-        public PipeReader Input { get; }
+        public PipeReader Input
+        {
+            get
+            {
+                if (_inputTask == null)
+                {
+                    RunAsync();
+                }
+
+                return _input.Reader;
+            }
+        }
 
         public PipeWriter Output
         {
             get
             {
                 if (_outputTask == null)
                 {
-                    _outputTask = WriteOutputAsync();
+                    RunAsync();
                 }
 
                 return _output.Writer;
             }
         }
 
-        public override ValueTask DisposeAsync()
+        public void RunAsync()
+        {
+            _inputTask = ReadInputAsync();
+            _outputTask = WriteOutputAsync();
+        }
+
+        public override async ValueTask DisposeAsync()
         {
             lock (_disposeLock)
             {
                 if (_disposed)
                 {
-                    return default;
+                    return;
                 }
                 _disposed = true;
             }
 
-            Input.Complete();
             _output.Writer.Complete();
+            _input.Reader.Complete();
 
-            if (_outputTask == null || _outputTask.IsCompletedSuccessfully)
+            if (_outputTask == null)
             {
-                // Wait for the output task to complete, this ensures that we've copied
-                // the application data to the underlying stream
-                return default;
+                return;
             }
 
-            return new ValueTask(_outputTask);
+            await _outputTask;
+
+            CancelPendingRead();
+
+            await _inputTask;
+        }
+
+        private async Task ReadInputAsync()
+        {
+            Exception error = null;
+            try
+            {
+                while (true)
+                {
+                    var outputBuffer = _input.Writer.GetMemory(_minAllocBufferSize);
+
+                    var bytesRead = await Stream.ReadAsync(outputBuffer);
+                    _input.Writer.Advance(bytesRead);
+
+                    if (bytesRead == 0)
+                    {
+                        // FIN
+                        break;
+                    }
+
+                    var result = await _input.Writer.FlushAsync();
```

Comment:
Is it possible that you could be stuck awaiting this call to flush indefinitely when DisposeAsync is called? Do we need to call CancelPendingFlush in the _input.Writer?

Reply:
Nvm you completed the _input.Reader which will unblock this call to FlushAsync()